examples/siamese_network/siamese_network.py

```diff
@@ -0,0 +1,322 @@
+
+
+class MatcherDataset(Dataset):
+    # following class implements data downloading and handles preprocessing
+    def __init__(self, root, train, download=False):
+        super(MatcherDataset, self).__init__()
+
+        # get CIFAR10 dataset
+        self.dataset = datasets.CIFAR10(root, train=train, download=download)
+
+        # convert data from numpy array to Tensor
+        self.data = torch.from_numpy(self.dataset.data)
+
+        # shift the dimensions of dataset to match the initial input layer dimensions
+        self.data = torch.movedim(self.data, (0, 1, 2, 3), (0, 2, 3, 1))
+
+        # convert targets list to torch Tensor
+        self.dataset.targets = torch.tensor(self.dataset.targets)
+
+        self.group_examples()
+
+    def group_examples(self):
+        """
+        To ease the accessibility of data based on the class, we will use `group_examples` to group
+        examples based on class. The data classes have already been mapped to numeric values and
+        so are the target outputs for each training input
+
+        Every key in `grouped_examples` corresponds to a class in CIFAR10 dataset. For every key in
+        `grouped_examples`, every value will conform to all of the indices for the CIFAR10
+        dataset examples that correspond to that key.
+        """
+
+        # get the targets from CIFAR10 dataset
+        np_arr = np.array(self.dataset.targets)
+
+        # group examples based on class
+        self.grouped_examples = {}
+        for i in range(0, 10):
+            self.grouped_examples[i] = np.where((np_arr == i))[0]
+
+    def __len__(self):
+        return self.data.shape[0]
+
+    def __getitem__(self, index):
+        """
+        For every sample in the batch we select 3 images. First one is the anchor image
+        which is the image obtained from the current index. We also obtain the label of
+        anchor image.
+
+        Now we select two random images, one belonging to the same class as that of the
+        anchor image (named as positive_image) and the other belonging to a different class
+        than that of the anchor image (named as negative_image). We return the anchor image,
+        positive image, negative image and anchor label.
+        """
+
+        # obtain the anchor image
+        anchor_image = self.data[index].float()
+
+        # obtain the class label of the anchor image
+        anchor_label = self.dataset.targets[index]
+        anchor_label = int(anchor_label.item())
+
+        # find a label which is different from anchor_label
+        neg_label = torch.randint(0, 10, (1,)).item()
+        while neg_label == anchor_label:
+            neg_label = random.randint(0, 9)
+
+        # find the index of image with positive label
+        positive_index = random.choice(self.grouped_examples[anchor_label])
+
+        # find the index of image with negative label
+        negative_index = random.choice(self.grouped_examples[neg_label])
```

Comment:
There is no alternative to random.choice as of now, according to the discussion [here](https://discuss.pytorch.org/t/torch-equivalent-of-numpy-random-choice/16146). An alternate way to implement it can be to use `torch.randint` in the range of incides of grouped_examples to generate a random index and choose a value using this index. Implementation:

```
random_index = torch.randint(0, len(self.grouped_examples[anchor_label]), (1, )).item()
positive_index = self.grouped_examples[anchor_label][random_index]
positive_image = self.data[positive_index].float()
random_index = torch.randint(0, len(self.grouped_examples[neg_label]), (1, )).item()
negative_index = self.grouped_examples[neg_label][random_index]
negative_image = self.data[negative_index].float()
```